### PR TITLE
Incorporate the OrderManager package.

### PR DESCRIPTION
**Pull request by mattmoor**
This is the key logic for handling the certificate ordering process.  The logic is based heavily on Go's autocert package, but reoriented into a form that is more amenable to a reconciliation loop.

I have not found any good fakes for the acme package that would allow for this package to be reasonably unit tested, so instead I wrote a very simple standalone sample that can be used to exercise it's functionality in a manner similar to autocert.

I have stood this up temporarily [here](https://foo.docs-on-the-rocks.io) to confirm that nothing broke in the copy.

WIP until https://github.com/knative/net-http01/pull/2 merges

### CODE REVIEW

File: cmd/sample/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+   ```shell
+   ko apply -Bf cmd/sample/sample.yaml
+   ```
+
+1. Curl the application:
+
+   ```shell
+   watch curl https://your-domain-name.io
+   ```
```
Comment (mattmoor) [inline cmd/sample/README.md]: Add trailing newline:
```suggestion
   ```

```

File: cmd/sample/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# net-http01 sample
+
+This small sample can be used to test the core libraries without all of the rest of knative.dev/serving.
```
Comment (mattmoor) [inline cmd/sample/README.md]: Format markdown:
```suggestion
This small sample can be used to test the core libraries without all of the rest
of knative.dev/serving.
```

File: cmd/sample/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+Instructions:
```
Comment (mattmoor) [inline cmd/sample/README.md]: Format markdown:
```suggestion
Instructions:

```

File: cmd/sample/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+1. Deploy the application.
+
+   ```shell
+   ko apply -Bf cmd/sample/sample.yaml
+   ```
+
+1. Curl the application:
+
+   ```shell
+   watch curl https://your-domain-name.io
```
Comment (mattmoor) [inline cmd/sample/README.md]: Format markdown:
```suggestion
   watch curl https://your-domain-name.io
   ```
```

File: pkg/ordermanager/ordermanager.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+	if t.err != nil {
+		om.cancelOrder(ctx, domains)
+		logger.Infof("Cancelling order for %v due to error: %v", domains, t.err)
+		return nil, nil, t.err
+	}
+
+	// See if the order specified by this ticket is ready.
+	status, err := t.GetStatus(ctx, om.Client)
+	if err != nil {
+		// TODO(mattmoor): If the error isn't transient,
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: File an issue and link?

File: pkg/ordermanager/ordermanager.go [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+	Endpoint = Production
+)
+
+// New creates a new OrderManager.
+func New(ctx context.Context, cb OrderUpCallback, chlr challenger.Interface) (Interface, error) {
+	acctKey, err := ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	a := &acme.Account{Contact: []string{}}
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: This is same as leaving Contact nil?

File: pkg/ordermanager/ordermanager.go [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+	client := &acme.Client{
+		DirectoryURL: Endpoint,
+		UserAgent:    "knative.dev/net-http01",
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: Extract to constant?

File: pkg/ordermanager/ordermanager.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+		Key:          acctKey,
+	}
+	_, err = client.Register(ctx, a, autocert.AcceptTOS)
+	if err != nil && err != acme.ErrAccountAlreadyExists {
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: Fold to previous line?

File: pkg/ordermanager/ordermanager.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+// Order implements Interface
+func (om *impl) Order(ctx context.Context, domains []string, owner interface{}) ([]*apis.URL, *tls.Certificate, error) {
+	logger := logging.FromContext(ctx)
+	t, ok := om.getTicket(domains)
+	if !ok {
+		// If there isn't an in-flight order, then initiate a new order.
+		var err error
+		t, err = om.initiateNewOrder(ctx, domains, owner)
+		if err != nil {
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: Fold to previous line?

File: pkg/ordermanager/ordermanager.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+		// then we should clear the ticket here, otherwise
+		// we are in an unrecoverable state.
+		return nil, nil, err
+	}
+	switch status {
+	case acme.StatusReady, acme.StatusValid:
+		logger.Infof("Order is ready for %v", domains)
+		// This removes the ticket, a subsequent Order will start
+		// the process over.
+		// TODO(mattmoor): Consider keeping completed orders around
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: File an issue and link?

File: pkg/ordermanager/ordermanager.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+		cert, err := om.completeOrder(ctx, domains, t)
+		return nil, cert, err
+
+	case acme.StatusPending, acme.StatusProcessing, acme.StatusUnknown:
+		logger.Infof("Order is pending for %v", domains)
+		urls, err := t.ChallengeURLs(ctx, om.Client)
+		return urls, nil, err
+
+	case acme.StatusDeactivated, acme.StatusExpired, acme.StatusInvalid,
+		acme.StatusRevoked:
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: why break to newline?

File: pkg/ordermanager/ordermanager.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+		// and return an error to the client which can retry as it sees
+		// fit.
+		om.cancelOrder(ctx, domains)
+		if err1, err2 := t.GetError(ctx, om.Client); err2 != nil {
+			// An error getting the error.
+			logging.FromContext(ctx).Errorf("Error getting error: %v", err)
+			return nil, nil, err2
+		} else if err1 != nil {
+			// The error returned by the CA leading to the above state.
+			logging.FromContext(ctx).Errorf("Error from theh CA: %v", err)
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: Typo: theh

File: pkg/ordermanager/ordermanager.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+			return nil, nil, err1
+		}
+		// Fallback on reporting the status.
+		logging.FromContext(ctx).Errorf("Bad status for order: %s", status)
+		return nil, nil, fmt.Errorf("Order resulted in status: %q", status)
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: s/status/bad status/ ?

File: pkg/ordermanager/ordermanager.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+type ticket struct {
+	uri   string
+	owner interface{}
+	err   error
+}
+
+// Order implements Interface
+func (om *impl) Order(ctx context.Context, domains []string, owner interface{}) ([]*apis.URL, *tls.Certificate, error) {
+	logger := logging.FromContext(ctx)
+	t, ok := om.getTicket(domains)
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: s/ok/found/ ?

File: pkg/ordermanager/ordermanager.go [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+
+func (t *ticket) GetStatus(ctx context.Context, client *acme.Client) (string, error) {
+	o, err := client.GetOrder(ctx, t.uri)
+	if err != nil {
+		return "", err
+	}
+	return o.Status, nil
+}
+
+func (t *ticket) GetError(ctx context.Context, client *acme.Client) (error, error) {
```
Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: s/GetError/GetErrors/?

Comment (tcnghia) [inline pkg/ordermanager/ordermanager.go]: name the output vars to distinguish between the two types of errors?

File: pkg/ordermanager/ordermanager.go (hunk 6 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: Done

File: pkg/ordermanager/ordermanager.go (hunk 7 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: Made it a var so folks can tweak.

File: pkg/ordermanager/ordermanager.go (hunk 8 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: done

File: pkg/ordermanager/ordermanager.go (hunk 14 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: done

File: pkg/ordermanager/ordermanager.go (hunk 9 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: done

File: pkg/ordermanager/ordermanager.go (hunk 5 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: For some of these cases it wasn't clear to me under what circumstances the client would return the error, so the TODOs are mostly to record places where I thought there was potential for issue, but without more information I'm not sure it's actionable.

File: pkg/ordermanager/ordermanager.go (hunk 10 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: TBH, I'm not sure it is worth the complexity.  I'll remove it.

File: pkg/ordermanager/ordermanager.go (hunk 11 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: unwrapped

File: pkg/ordermanager/ordermanager.go (hunk 12 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: fixed

File: pkg/ordermanager/ordermanager.go (hunk 13 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: done

File: pkg/ordermanager/ordermanager.go (hunk 15 above)
Comment (mattmoor) [reply pkg/ordermanager/ordermanager.go]: done

File: pkg/ordermanager/ordermanager.go [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+	// UserAgent is the HTTP user agent that is used with the ACME client,
+	// so that traffic from this client may be distinguished from others.
+	// It is intentionally exposed as a mutable variable in case users with
+	// to distinguish their traffic further.
+	UserAgent = "knative.dev/net-http01"
+)
+
+// New creates a new OrderManager.
+func New(ctx context.Context, cb OrderUpCallback, chlr challenger.Interface) (Interface, error) {
+	acctKey, err := ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
```
Comment (ZhiminXiang) [inline pkg/ordermanager/ordermanager.go]: we can also store the account key in a secret.
So when the controller is restarted, it does not register a new account which won't break the order that is being processed with the old account.

File: pkg/ordermanager/ordermanager.go [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+
+	case acme.StatusDeactivated, acme.StatusExpired, acme.StatusInvalid, acme.StatusRevoked:
+		logger.Infof("Order is invalid for %v", domains)
+		// This is a permanently bad state, we should flush the ticket
+		// and return an error to the client which can retry as it sees
+		// fit.
+		om.cancelOrder(ctx, domains)
+		if err1, err2 := t.GetError(ctx, om.Client); err2 != nil {
+			// An error getting the error.
+			logging.FromContext(ctx).Errorf("Error getting error: %v", err)
```
Comment (ZhiminXiang) [inline pkg/ordermanager/ordermanager.go]: Should this be err2?

File: pkg/ordermanager/ordermanager.go [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+			return nil, nil, err2
+		} else if err1 != nil {
+			// The error returned by the CA leading to the above state.
+			logging.FromContext(ctx).Errorf("Error from the CA: %v", err)
```
Comment (ZhiminXiang) [inline pkg/ordermanager/ordermanager.go]: Should this be err1?

**Pull request merged**